src/stdlib_str2num.fypp

```diff
@@ -0,0 +1,553 @@
+!>   difference abs :                                 0.4625E-475 
+!>   difference rel :                                 0.3300E-029%
+
+module stdlib_str2num
+    use stdlib_kinds, only: sp, dp, qp, int8, int16, int32, int64
+    use ieee_arithmetic
+    implicit none
+    private
+    public :: to_num, to_num_p
+    
+    integer, parameter :: ikind = selected_int_kind(2)
+    integer(kind=ikind), parameter :: digit_0    = ichar('0',kind=ikind)
+    integer(kind=ikind), parameter :: period     = ichar('.',kind=ikind) - digit_0
+    integer(kind=ikind), parameter :: comma      = ichar(',',kind=ikind) - digit_0
+    integer(kind=ikind), parameter :: minus_sign = ichar('-',kind=ikind) - digit_0
+    integer(kind=ikind), parameter :: plus_sign  = ichar('+',kind=ikind) - digit_0
+    integer(kind=ikind), parameter :: Inf        = ichar('I',kind=ikind) 
+    integer(kind=ikind), parameter :: NaN        = ichar('N',kind=ikind) 
+    integer(kind=ikind), parameter :: le         = ichar('e',kind=ikind) - digit_0
+    integer(kind=ikind), parameter :: BE         = ichar('E',kind=ikind) - digit_0
+    integer(kind=ikind), parameter :: ld         = ichar('d',kind=ikind) - digit_0
+    integer(kind=ikind), parameter :: BD         = ichar('D',kind=ikind) - digit_0
+    integer(kind=ikind), parameter :: LF = 10, CR = 13, WS = 32
+
+    interface to_num
+        #:for k1, t1 in INT_KINDS_TYPES
+        module procedure to_${k1}$
+        #:endfor
+        module procedure to_float
+        module procedure to_double
+#:if WITH_QP
+        module procedure to_quad
+#:endif
+    end interface
+
+    interface to_num_p
+        #:for k1, t1 in INT_KINDS_TYPES
+        module procedure to_${k1}$_p
+        #:endfor
+        module procedure to_float_p
+        module procedure to_double_p
+#:if WITH_QP
+        module procedure to_quad_p
+#:endif
+    end interface
+
+    interface to_num_base
+        #:for k1, t1 in INT_KINDS_TYPES
+        module procedure to_int_${k1}$
+        #:endfor
+        module procedure to_real_sp
+        module procedure to_real_dp
+#:if WITH_QP
+        module procedure to_real_qp
+#:endif
+    end interface
+    
+    contains
+    
+    !---------------------------------------------
+    ! String To Number interfaces
+    !---------------------------------------------
+
+    #:for k1, t1 in INT_KINDS_TYPES
+    elemental function to_${k1}$(s,mold) result(v)
+        ! -- In/out Variables
+        character(*), intent(in) :: s !> input string
+        ${t1}$, intent(in) :: mold !> dummy argument to disambiguate at compile time the generic interface
+        ${t1}$ :: v !> Output integer 32 value
+        ! -- Internal Variables
+        integer(int8) :: p !> position within the number
+        integer(int8)  :: stat !> error status
+        !----------------------------------------------
+        call to_num_base(s,v,p,stat)
+    end function
+    
+    function to_${k1}$_p(s,mold,stat) result(v)
```

Comment:
could this function be pure?

Reply:
It can't because of this `s => s(p:)` where the pointer is shifted from within.